{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4169272",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/evsi8432/opt/anaconda3/lib/python3.9/site-packages/scipy/__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.26.4\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import autograd.numpy as np\n",
    "from autograd.scipy.special import logsumexp,erf\n",
    "from autograd.scipy.stats import norm\n",
    "from autograd import grad\n",
    "\n",
    "import os\n",
    "#from autograd_gamma import gamma, gammainc, gammaincc, gammaincln, gammainccln\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from copy import deepcopy\n",
    "from itertools import product\n",
    "import pickle\n",
    "\n",
    "from tree import Tree\n",
    "from optimizer import Optimizer\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dfea8f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = \"DS14\"\n",
    "\n",
    "# load in data from fasta\n",
    "data = {}\n",
    "filename = \"../dat/\"+ds+\"/\"+ds+\".fasta\"\n",
    "with open(filename, 'r') as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\">\"):\n",
    "            species = line.replace(\">\",\"\").replace(\"\\n\",\"\")\n",
    "            data[species] = \"\"\n",
    "        else:\n",
    "            data[species] += line.replace(\">\",\"\").replace(\"\\n\",\"\").replace(\".\",\"-\")\n",
    "\n",
    "\n",
    "n_tax = len(data)\n",
    "n_char = len(data[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a6d0fb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa = list(data.keys())\n",
    "sub_data = {}\n",
    "for sub_size in [48,24,12,6,3]:\n",
    "    taxa = np.random.choice(taxa,size=sub_size,replace=False)\n",
    "    sub_data[sub_size] = {taxon: data[taxon] for taxon in taxa}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "269ddd6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to pickle file\n",
    "for N in [48,24,12,6,3]:\n",
    "    if not os.path.exists(\"../dat/DS14_\" + str(N)):\n",
    "        os.makedirs(\"../dat/DS14_\" + str(N))\n",
    "    filename = \"../dat/DS14_\" + str(N) +\"/DS14_\" + str(N) + \".pickle\"\n",
    "    with open(filename, 'wb') as handle:\n",
    "        pickle.dump(sub_data[N], handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "989687da",
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 3/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7804bfed",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "96",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [31]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m N \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(N\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m2\u001b[39m)\n\u001b[1;32m      2\u001b[0m ds \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDS14_\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m%\u001b[39mN\n\u001b[0;32m----> 3\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[43msub_data\u001b[49m\u001b[43m[\u001b[49m\u001b[43mN\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      4\u001b[0m n_tax \u001b[38;5;241m=\u001b[39m N\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(N)\n",
      "\u001b[0;31mKeyError\u001b[0m: 96"
     ]
    }
   ],
   "source": [
    "N = int(N*2)\n",
    "ds = \"DS14_%d\"%N\n",
    "data = sub_data[N]\n",
    "n_tax = N\n",
    "print(N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d659455f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write nexus file\n",
    "filename = \"../dat/\"+ds+\"/\"+ds+\".nexus\"\n",
    "f = open(filename, \"w\")\n",
    "f.write(\"\"\"#NEXUS\n",
    "\n",
    "BEGIN DATA;\n",
    "\\tDIMENSIONS NTAX=%d NCHAR=%d;\n",
    "\\tFORMAT MISSING=? GAP=- DATATYPE=DNA;\n",
    "\\tMATRIX\n",
    "\"\"\" % (n_tax,n_char))\n",
    "for key in data:\n",
    "    f.write(\"\\t\")\n",
    "    f.write(key.replace(\"-\",\"_\"))\n",
    "    f.write(\"\\t\")\n",
    "    f.write(data[key])\n",
    "    f.write(\"\\n\")\n",
    "\n",
    "f.write(\"\"\"\\t;\n",
    "END;\"\"\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "78d9cfe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make xml file for BEAST candidate runs\n",
    "\n",
    "filename = \"../dat/\"+ds+\"/\"+ds+\".xml\"\n",
    "f = open(filename, \"w\")\n",
    "f.write('''<?xml version=\"1.0\" standalone=\"yes\"?>\n",
    "\n",
    "<!-- Generated by BEAUTi v1.10.4 Prerelease #bc6cbd9                         -->\n",
    "<!--       by Alexei J. Drummond, Andrew Rambaut and Marc A. Suchard         -->\n",
    "<!--       Department of Computer Science, University of Auckland and        -->\n",
    "<!--       Institute of Evolutionary Biology, University of Edinburgh        -->\n",
    "<!--       David Geffen School of Medicine, University of California, Los Angeles-->\n",
    "<!--       http://beast.community/                                           -->\n",
    "<beast version=\"1.10.4\">\n",
    "\t\n",
    "\n",
    "\t<!-- The list of taxa to be analysed (can also include dates/ages).          -->\n",
    "\t<!-- ntax=%d                                                                 -->\n",
    "\t<taxa id=\"taxa\">\n",
    "''' % n_tax)\n",
    "\n",
    "for species in data:\n",
    "    f.write('\\t\\t<taxon id=\"%s\"/> \\n' % species)\n",
    "\n",
    "f.write('''\\t</taxa>\n",
    "\t\n",
    "\t\n",
    "\n",
    "\t<!-- The sequence alignment (each sequence refers to a taxon above).         -->\n",
    "\t<!-- ntax=%d nchar=%d                                                      -->\n",
    "\t<alignment id=\"alignment\" dataType=\"nucleotide\">\n",
    "''' % (n_tax,n_char))\n",
    "\n",
    "for species in data:\n",
    "    f.write(\"\\t\\t<sequence>\\n\")\n",
    "    f.write('\\t\\t\\t<taxon idref=\"%s\"/>\\n'%species)\n",
    "    f.write('\\t\\t\\t%s\\n'%data[species])\n",
    "    f.write(\"\\t\\t</sequence>\\n\")\n",
    "f.write(\"\\t</alignment>\\n\")\n",
    "\n",
    "f.write(\"\"\"\n",
    "\n",
    "\t<!-- The unique patterns from 1 to end                                       -->\n",
    "\t<!-- npatterns=933                                                           -->\n",
    "\t<patterns id=\"patterns\" from=\"1\" strip=\"false\">\n",
    "\t\t<alignment idref=\"alignment\"/>\n",
    "\t</patterns>\n",
    "\t\n",
    "\n",
    "\t<!-- A prior assumption that the population size has remained constant       -->\n",
    "\t<!-- throughout the time spanned by the genealogy.                           -->\n",
    "\t<constantSize id=\"constant\" units=\"years\">\n",
    "\t\t<populationSize>\n",
    "\t\t\t<parameter id=\"constant.popSize\" value=\"5.0\"/>\n",
    "\t\t</populationSize>\n",
    "\t</constantSize>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a random starting tree under the coalescent process            -->\n",
    "\t<coalescentSimulator id=\"startingTree\">\n",
    "\t\t<taxa idref=\"taxa\"/>\n",
    "\t\t<constantSize idref=\"constant\"/>\n",
    "\t</coalescentSimulator>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a tree model                                                   -->\n",
    "\t<treeModel id=\"treeModel\">\n",
    "\t\t<coalescentTree idref=\"startingTree\"/>\n",
    "\t\t<rootHeight>\n",
    "\t\t\t<parameter id=\"treeModel.rootHeight\"/>\n",
    "\t\t</rootHeight>\n",
    "\t\t<nodeHeights internalNodes=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t\t<nodeHeights internalNodes=\"true\" rootNode=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t</treeModel>\n",
    "\n",
    "\t<!-- Statistic for sum of the branch lengths of the tree (tree length)       -->\n",
    "\t<treeLengthStatistic id=\"treeLength\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t</treeLengthStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a coalescent likelihood                                        -->\n",
    "\t<coalescentLikelihood id=\"coalescent\">\n",
    "\t\t<model>\n",
    "\t\t\t<constantSize idref=\"constant\"/>\n",
    "\t\t</model>\n",
    "\t\t<populationTree>\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</populationTree>\n",
    "\t</coalescentLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- The strict clock (Uniform rates across branches)                        -->\n",
    "\t<strictClockBranchRates id=\"branchRates\">\n",
    "\t\t<rate>\n",
    "\t\t\t<parameter id=\"clock.rate\" value=\"1.0\"/>\n",
    "\t\t</rate>\n",
    "\t</strictClockBranchRates>\n",
    "\t\n",
    "\t<rateStatistic id=\"meanRate\" name=\"meanRate\" mode=\"mean\" internal=\"true\" external=\"true\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</rateStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- The JC substitution model (Jukes & Cantor, 1969)                        -->\n",
    "\t<HKYModel id=\"jc\">\n",
    "\t\t<frequencies>\n",
    "\t\t\t<frequencyModel dataType=\"nucleotide\">\n",
    "\t\t\t\t<frequencies>\n",
    "\t\t\t\t\t<parameter id=\"frequencies\" value=\"0.25 0.25 0.25 0.25\"/>\n",
    "\t\t\t\t</frequencies>\n",
    "\t\t\t</frequencyModel>\n",
    "\t\t</frequencies>\n",
    "\t\t<kappa>\n",
    "\t\t\t<parameter value=\"1.0\"/>\n",
    "\t\t</kappa>\n",
    "\t</HKYModel>\n",
    "\n",
    "\t<!-- site model                                                              -->\n",
    "\t<siteModel id=\"siteModel\">\n",
    "\t\t<substitutionModel>\n",
    "\t\t\t<HKYModel idref=\"jc\"/>\n",
    "\t\t</substitutionModel>\n",
    "\t</siteModel>\n",
    "\n",
    "\t<!--                                                                         -->\n",
    "\t<statistic id=\"mu\" name=\"mu\">\n",
    "\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t</statistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Likelihood for tree given sequence data                                 -->\n",
    "\t<treeDataLikelihood id=\"treeLikelihood\" useAmbiguities=\"false\">\n",
    "\t\t<partition>\n",
    "\t\t\t<patterns idref=\"patterns\"/>\n",
    "\t\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t\t</partition>\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</treeDataLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- Define operators                                                        -->\n",
    "\t<operators id=\"operators\" optimizationSchedule=\"default\">\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" scaleAll=\"true\" ignoreBounds=\"true\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<subtreeSlide size=\"1.0\" gaussian=\"true\" weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</subtreeSlide>\n",
    "\t\t<narrowExchange weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</narrowExchange>\n",
    "\t\t<wideExchange weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wideExchange>\n",
    "\t\t<wilsonBalding weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wilsonBalding>\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<uniformOperator weight=\"30\">\n",
    "\t\t\t<parameter idref=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</uniformOperator>\n",
    "\t</operators>\n",
    "\t\n",
    "\n",
    "\t<!-- Define MCMC                                                             -->\n",
    "\t<mcmc id=\"mcmc\" chainLength=\"1000000\" autoOptimize=\"true\" operatorAnalysis=\"%s.ops\">\n",
    "\t\t<joint id=\"joint\">\n",
    "\t\t\t<prior id=\"prior\">\n",
    "\t\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\t\n",
    "\t\t\t\t\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</prior>\n",
    "\t\t\t<likelihood id=\"likelihood\">\n",
    "\t\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t</likelihood>\n",
    "\t\t</joint>\n",
    "\t\t<operators idref=\"operators\"/>\n",
    "\n",
    "\t\t<!-- write log to screen                                                     -->\n",
    "\t\t<log id=\"screenLog\" logEvery=\"1000\">\n",
    "\t\t\t<column label=\"Joint\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Prior\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Likelihood\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"rootHeight\" sf=\"6\" width=\"12\">\n",
    "\t\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t</column>\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write log to file                                                       -->\n",
    "\t\t<log id=\"fileLog\" logEvery=\"1000\" fileName=\"%s.log\" overwrite=\"false\">\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t<treeLengthStatistic idref=\"treeLength\"/>\n",
    "\t\t\t<parameter idref=\"constant.popSize\"/>\n",
    "\t\t\t<parameter idref=\"clock.rate\"/>\n",
    "\t\t\t<rateStatistic idref=\"meanRate\"/>\n",
    "\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write tree log to file                                                  -->\n",
    "\t\t<logTree id=\"treeFileLog\" logEvery=\"1000\" nexusFormat=\"true\" fileName=\"%s.trees\" sortTranslationTable=\"true\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t\t<trait name=\"rate\" tag=\"rate\">\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</trait>\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t</logTree>\n",
    "\t</mcmc>\n",
    "\t\n",
    "\t<report>\n",
    "\t\t<property name=\"timer\">\n",
    "\t\t\t<mcmc idref=\"mcmc\"/>\n",
    "\t\t</property>\n",
    "\t</report>\n",
    "\t\n",
    "</beast>\n",
    "\"\"\" % (ds,ds,ds))\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b3c527e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make xml file for BEAST golden run\n",
    "\n",
    "filename = \"../dat/\"+ds+\"/\"+ds+\"_golden.xml\"\n",
    "f = open(filename, \"w\")\n",
    "f.write('''<?xml version=\"1.0\" standalone=\"yes\"?>\n",
    "\n",
    "<!-- Generated by BEAUTi v1.10.4 Prerelease #bc6cbd9                         -->\n",
    "<!--       by Alexei J. Drummond, Andrew Rambaut and Marc A. Suchard         -->\n",
    "<!--       Department of Computer Science, University of Auckland and        -->\n",
    "<!--       Institute of Evolutionary Biology, University of Edinburgh        -->\n",
    "<!--       David Geffen School of Medicine, University of California, Los Angeles-->\n",
    "<!--       http://beast.community/                                           -->\n",
    "<beast version=\"1.10.4\">\n",
    "\t\n",
    "\n",
    "\t<!-- The list of taxa to be analysed (can also include dates/ages).          -->\n",
    "\t<!-- ntax=%d                                                                 -->\n",
    "\t<taxa id=\"taxa\">\n",
    "''' % n_tax)\n",
    "\n",
    "for species in data:\n",
    "    f.write('\\t\\t<taxon id=\"%s\"/> \\n' % species)\n",
    "\n",
    "f.write('''\\t</taxa>\n",
    "\t\n",
    "\t\n",
    "\n",
    "\t<!-- The sequence alignment (each sequence refers to a taxon above).         -->\n",
    "\t<!-- ntax=%d nchar=%d                                                      -->\n",
    "\t<alignment id=\"alignment\" dataType=\"nucleotide\">\n",
    "''' % (n_tax,n_char))\n",
    "\n",
    "for species in data:\n",
    "    f.write(\"\\t\\t<sequence>\\n\")\n",
    "    f.write('\\t\\t\\t<taxon idref=\"%s\"/>\\n'%species)\n",
    "    f.write('\\t\\t\\t%s\\n'%data[species])\n",
    "    f.write(\"\\t\\t</sequence>\\n\")\n",
    "f.write(\"\\t</alignment>\\n\")\n",
    "\n",
    "f.write(\"\"\"\n",
    "\n",
    "\t<!-- The unique patterns from 1 to end                                       -->\n",
    "\t<!-- npatterns=933                                                           -->\n",
    "\t<patterns id=\"patterns\" from=\"1\" strip=\"false\">\n",
    "\t\t<alignment idref=\"alignment\"/>\n",
    "\t</patterns>\n",
    "\t\n",
    "\n",
    "\t<!-- A prior assumption that the population size has remained constant       -->\n",
    "\t<!-- throughout the time spanned by the genealogy.                           -->\n",
    "\t<constantSize id=\"constant\" units=\"years\">\n",
    "\t\t<populationSize>\n",
    "\t\t\t<parameter id=\"constant.popSize\" value=\"5.0\"/>\n",
    "\t\t</populationSize>\n",
    "\t</constantSize>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a random starting tree under the coalescent process            -->\n",
    "\t<coalescentSimulator id=\"startingTree\">\n",
    "\t\t<taxa idref=\"taxa\"/>\n",
    "\t\t<constantSize idref=\"constant\"/>\n",
    "\t</coalescentSimulator>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a tree model                                                   -->\n",
    "\t<treeModel id=\"treeModel\">\n",
    "\t\t<coalescentTree idref=\"startingTree\"/>\n",
    "\t\t<rootHeight>\n",
    "\t\t\t<parameter id=\"treeModel.rootHeight\"/>\n",
    "\t\t</rootHeight>\n",
    "\t\t<nodeHeights internalNodes=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t\t<nodeHeights internalNodes=\"true\" rootNode=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t</treeModel>\n",
    "\n",
    "\t<!-- Statistic for sum of the branch lengths of the tree (tree length)       -->\n",
    "\t<treeLengthStatistic id=\"treeLength\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t</treeLengthStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a coalescent likelihood                                        -->\n",
    "\t<coalescentLikelihood id=\"coalescent\">\n",
    "\t\t<model>\n",
    "\t\t\t<constantSize idref=\"constant\"/>\n",
    "\t\t</model>\n",
    "\t\t<populationTree>\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</populationTree>\n",
    "\t</coalescentLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- The strict clock (Uniform rates across branches)                        -->\n",
    "\t<strictClockBranchRates id=\"branchRates\">\n",
    "\t\t<rate>\n",
    "\t\t\t<parameter id=\"clock.rate\" value=\"1.0\"/>\n",
    "\t\t</rate>\n",
    "\t</strictClockBranchRates>\n",
    "\t\n",
    "\t<rateStatistic id=\"meanRate\" name=\"meanRate\" mode=\"mean\" internal=\"true\" external=\"true\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</rateStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- The JC substitution model (Jukes & Cantor, 1969)                        -->\n",
    "\t<HKYModel id=\"jc\">\n",
    "\t\t<frequencies>\n",
    "\t\t\t<frequencyModel dataType=\"nucleotide\">\n",
    "\t\t\t\t<frequencies>\n",
    "\t\t\t\t\t<parameter id=\"frequencies\" value=\"0.25 0.25 0.25 0.25\"/>\n",
    "\t\t\t\t</frequencies>\n",
    "\t\t\t</frequencyModel>\n",
    "\t\t</frequencies>\n",
    "\t\t<kappa>\n",
    "\t\t\t<parameter value=\"1.0\"/>\n",
    "\t\t</kappa>\n",
    "\t</HKYModel>\n",
    "\n",
    "\t<!-- site model                                                              -->\n",
    "\t<siteModel id=\"siteModel\">\n",
    "\t\t<substitutionModel>\n",
    "\t\t\t<HKYModel idref=\"jc\"/>\n",
    "\t\t</substitutionModel>\n",
    "\t</siteModel>\n",
    "\n",
    "\t<!--                                                                         -->\n",
    "\t<statistic id=\"mu\" name=\"mu\">\n",
    "\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t</statistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Likelihood for tree given sequence data                                 -->\n",
    "\t<treeDataLikelihood id=\"treeLikelihood\" useAmbiguities=\"false\">\n",
    "\t\t<partition>\n",
    "\t\t\t<patterns idref=\"patterns\"/>\n",
    "\t\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t\t</partition>\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</treeDataLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- Define operators                                                        -->\n",
    "\t<operators id=\"operators\" optimizationSchedule=\"default\">\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" scaleAll=\"true\" ignoreBounds=\"true\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<subtreeSlide size=\"1.0\" gaussian=\"true\" weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</subtreeSlide>\n",
    "\t\t<narrowExchange weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</narrowExchange>\n",
    "\t\t<wideExchange weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wideExchange>\n",
    "\t\t<wilsonBalding weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wilsonBalding>\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<uniformOperator weight=\"30\">\n",
    "\t\t\t<parameter idref=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</uniformOperator>\n",
    "\t</operators>\n",
    "\t\n",
    "\n",
    "\t<!-- Define MCMC                                                             -->\n",
    "\t<mcmc id=\"mcmc\" chainLength=\"100000000\" autoOptimize=\"true\" operatorAnalysis=\"%s.ops\">\n",
    "\t\t<joint id=\"joint\">\n",
    "\t\t\t<prior id=\"prior\">\n",
    "\t\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\t\n",
    "\t\t\t\t\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</prior>\n",
    "\t\t\t<likelihood id=\"likelihood\">\n",
    "\t\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t</likelihood>\n",
    "\t\t</joint>\n",
    "\t\t<operators idref=\"operators\"/>\n",
    "\n",
    "\t\t<!-- write log to screen                                                     -->\n",
    "\t\t<log id=\"screenLog\" logEvery=\"10000\">\n",
    "\t\t\t<column label=\"Joint\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Prior\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Likelihood\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"rootHeight\" sf=\"6\" width=\"12\">\n",
    "\t\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t</column>\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write log to file                                                       -->\n",
    "\t\t<log id=\"fileLog\" logEvery=\"10000\" fileName=\"%s.log\" overwrite=\"false\">\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t<treeLengthStatistic idref=\"treeLength\"/>\n",
    "\t\t\t<parameter idref=\"constant.popSize\"/>\n",
    "\t\t\t<parameter idref=\"clock.rate\"/>\n",
    "\t\t\t<rateStatistic idref=\"meanRate\"/>\n",
    "\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write tree log to file                                                  -->\n",
    "\t\t<logTree id=\"treeFileLog\" logEvery=\"10000\" nexusFormat=\"true\" fileName=\"%s.trees\" sortTranslationTable=\"true\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t\t<trait name=\"rate\" tag=\"rate\">\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</trait>\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t</logTree>\n",
    "\t</mcmc>\n",
    "\n",
    "\t<report>\n",
    "\t\t<property name=\"timer\">\n",
    "\t\t\t<mcmc idref=\"mcmc\"/>\n",
    "\t\t</property>\n",
    "\t</report>\n",
    "\t\n",
    "</beast>\n",
    "\"\"\" % (ds,ds,ds))\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "65b71152",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make xml file for BEAST SS analysis for marginal likelihood\n",
    "\n",
    "filename = \"../dat/\"+ds+\"/\"+ds+\"_MLL.xml\"\n",
    "f = open(filename, \"w\")\n",
    "f.write('''<?xml version=\"1.0\" standalone=\"yes\"?>\n",
    "\n",
    "<!-- Generated by BEAUTi v1.10.4 Prerelease #bc6cbd9                         -->\n",
    "<!--       by Alexei J. Drummond, Andrew Rambaut and Marc A. Suchard         -->\n",
    "<!--       Department of Computer Science, University of Auckland and        -->\n",
    "<!--       Institute of Evolutionary Biology, University of Edinburgh        -->\n",
    "<!--       David Geffen School of Medicine, University of California, Los Angeles-->\n",
    "<!--       http://beast.community/                                           -->\n",
    "<beast version=\"1.10.4\">\n",
    "\t\n",
    "\n",
    "\t<!-- The list of taxa to be analysed (can also include dates/ages).          -->\n",
    "\t<!-- ntax=%d                                                                 -->\n",
    "\t<taxa id=\"taxa\">\n",
    "''' % n_tax)\n",
    "\n",
    "for species in data:\n",
    "    f.write('\\t\\t<taxon id=\"%s\"/> \\n' % species)\n",
    "\n",
    "f.write('''\\t</taxa>\n",
    "\t\n",
    "\t\n",
    "\n",
    "\t<!-- The sequence alignment (each sequence refers to a taxon above).         -->\n",
    "\t<!-- ntax=%d nchar=%d                                                      -->\n",
    "\t<alignment id=\"alignment\" dataType=\"nucleotide\">\n",
    "''' % (n_tax,n_char))\n",
    "\n",
    "for species in data:\n",
    "    f.write(\"\\t\\t<sequence>\\n\")\n",
    "    f.write('\\t\\t\\t<taxon idref=\"%s\"/>\\n'%species)\n",
    "    f.write('\\t\\t\\t%s\\n'%data[species])\n",
    "    f.write(\"\\t\\t</sequence>\\n\")\n",
    "f.write(\"\\t</alignment>\\n\")\n",
    "\n",
    "f.write(\"\"\"\n",
    "\n",
    "\t<!-- The unique patterns from 1 to end                                       -->\n",
    "\t<patterns id=\"patterns\" from=\"1\" strip=\"false\">\n",
    "\t\t<alignment idref=\"alignment\"/>\n",
    "\t</patterns>\n",
    "\t\n",
    "\n",
    "\t<!-- A prior assumption that the population size has remained constant       -->\n",
    "\t<!-- throughout the time spanned by the genealogy.                           -->\n",
    "\t<constantSize id=\"constant\" units=\"years\">\n",
    "\t\t<populationSize>\n",
    "\t\t\t<parameter id=\"constant.popSize\" value=\"5.0\"/>\n",
    "\t\t</populationSize>\n",
    "\t</constantSize>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a random starting tree under the coalescent process            -->\n",
    "\t<coalescentSimulator id=\"startingTree\">\n",
    "\t\t<taxa idref=\"taxa\"/>\n",
    "\t\t<constantSize idref=\"constant\"/>\n",
    "\t</coalescentSimulator>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a tree model                                                   -->\n",
    "\t<treeModel id=\"treeModel\">\n",
    "\t\t<coalescentTree idref=\"startingTree\"/>\n",
    "\t\t<rootHeight>\n",
    "\t\t\t<parameter id=\"treeModel.rootHeight\"/>\n",
    "\t\t</rootHeight>\n",
    "\t\t<nodeHeights internalNodes=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t\t<nodeHeights internalNodes=\"true\" rootNode=\"true\">\n",
    "\t\t\t<parameter id=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</nodeHeights>\n",
    "\t</treeModel>\n",
    "\n",
    "\t<!-- Statistic for sum of the branch lengths of the tree (tree length)       -->\n",
    "\t<treeLengthStatistic id=\"treeLength\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t</treeLengthStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Generate a coalescent likelihood                                        -->\n",
    "\t<coalescentLikelihood id=\"coalescent\">\n",
    "\t\t<model>\n",
    "\t\t\t<constantSize idref=\"constant\"/>\n",
    "\t\t</model>\n",
    "\t\t<populationTree>\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</populationTree>\n",
    "\t</coalescentLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- The strict clock (Uniform rates across branches)                        -->\n",
    "\t<strictClockBranchRates id=\"branchRates\">\n",
    "\t\t<rate>\n",
    "\t\t\t<parameter id=\"clock.rate\" value=\"1.0\"/>\n",
    "\t\t</rate>\n",
    "\t</strictClockBranchRates>\n",
    "\t\n",
    "\t<rateStatistic id=\"meanRate\" name=\"meanRate\" mode=\"mean\" internal=\"true\" external=\"true\">\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</rateStatistic>\n",
    "\t\n",
    "\n",
    "\t<!-- The JC substitution model (Jukes & Cantor, 1969)                        -->\n",
    "\t<HKYModel id=\"jc\">\n",
    "\t\t<frequencies>\n",
    "\t\t\t<frequencyModel dataType=\"nucleotide\">\n",
    "\t\t\t\t<frequencies>\n",
    "\t\t\t\t\t<parameter id=\"frequencies\" value=\"0.25 0.25 0.25 0.25\"/>\n",
    "\t\t\t\t</frequencies>\n",
    "\t\t\t</frequencyModel>\n",
    "\t\t</frequencies>\n",
    "\t\t<kappa>\n",
    "\t\t\t<parameter value=\"1.0\"/>\n",
    "\t\t</kappa>\n",
    "\t</HKYModel>\n",
    "\n",
    "\t<!-- site model                                                              -->\n",
    "\t<siteModel id=\"siteModel\">\n",
    "\t\t<substitutionModel>\n",
    "\t\t\t<HKYModel idref=\"jc\"/>\n",
    "\t\t</substitutionModel>\n",
    "\t</siteModel>\n",
    "\n",
    "\t<!--                                                                         -->\n",
    "\t<statistic id=\"mu\" name=\"mu\">\n",
    "\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t</statistic>\n",
    "\t\n",
    "\n",
    "\t<!-- Likelihood for tree given sequence data                                 -->\n",
    "\t<treeDataLikelihood id=\"treeLikelihood\" useAmbiguities=\"false\">\n",
    "\t\t<partition>\n",
    "\t\t\t<patterns idref=\"patterns\"/>\n",
    "\t\t\t<siteModel idref=\"siteModel\"/>\n",
    "\t\t</partition>\n",
    "\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t</treeDataLikelihood>\n",
    "\t\n",
    "\n",
    "\t<!-- Define operators                                                        -->\n",
    "\t<operators id=\"operators\" optimizationSchedule=\"default\">\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" scaleAll=\"true\" ignoreBounds=\"true\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.allInternalNodeHeights\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<subtreeSlide size=\"1.0\" gaussian=\"true\" weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</subtreeSlide>\n",
    "\t\t<narrowExchange weight=\"30\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</narrowExchange>\n",
    "\t\t<wideExchange weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wideExchange>\n",
    "\t\t<wilsonBalding weight=\"3\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t</wilsonBalding>\n",
    "\t\t<scaleOperator scaleFactor=\"0.75\" weight=\"3\">\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t</scaleOperator>\n",
    "\t\t<uniformOperator weight=\"30\">\n",
    "\t\t\t<parameter idref=\"treeModel.internalNodeHeights\"/>\n",
    "\t\t</uniformOperator>\n",
    "\t</operators>\n",
    "\t\n",
    "\n",
    "\t<!-- Define MCMC                                                             -->\n",
    "\t<mcmc id=\"mcmc\" chainLength=\"10000000\" autoOptimize=\"true\" operatorAnalysis=\"%s.ops\">\n",
    "\t\t<joint id=\"joint\">\n",
    "\t\t\t<prior id=\"prior\">\n",
    "\t\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\t\n",
    "\t\t\t\t\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</prior>\n",
    "\t\t\t<likelihood id=\"likelihood\">\n",
    "\t\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t</likelihood>\n",
    "\t\t</joint>\n",
    "\t\t<operators idref=\"operators\"/>\n",
    "\n",
    "\t\t<!-- write log to screen                                                     -->\n",
    "\t\t<log id=\"screenLog\" logEvery=\"1000\">\n",
    "\t\t\t<column label=\"Joint\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Prior\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"Likelihood\" dp=\"4\" width=\"12\">\n",
    "\t\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t</column>\n",
    "\t\t\t<column label=\"rootHeight\" sf=\"6\" width=\"12\">\n",
    "\t\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t</column>\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write log to file                                                       -->\n",
    "\t\t<log id=\"fileLog\" logEvery=\"1000\" fileName=\"%s.log\" overwrite=\"false\">\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t<likelihood idref=\"likelihood\"/>\n",
    "\t\t\t<parameter idref=\"treeModel.rootHeight\"/>\n",
    "\t\t\t<treeLengthStatistic idref=\"treeLength\"/>\n",
    "\t\t\t<parameter idref=\"constant.popSize\"/>\n",
    "\t\t\t<parameter idref=\"clock.rate\"/>\n",
    "\t\t\t<rateStatistic idref=\"meanRate\"/>\n",
    "\t\t\t<treeDataLikelihood idref=\"treeLikelihood\"/>\n",
    "\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t<coalescentLikelihood idref=\"coalescent\"/>\n",
    "\t\t\t\n",
    "\t\t</log>\n",
    "\n",
    "\t\t<!-- write tree log to file                                                  -->\n",
    "\t\t<logTree id=\"treeFileLog\" logEvery=\"1000\" nexusFormat=\"true\" fileName=\"%s.trees\" sortTranslationTable=\"true\">\n",
    "\t\t\t<treeModel idref=\"treeModel\"/>\n",
    "\t\t\t<trait name=\"rate\" tag=\"rate\">\n",
    "\t\t\t\t<strictClockBranchRates idref=\"branchRates\"/>\n",
    "\t\t\t</trait>\n",
    "\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t</logTree>\n",
    "\t</mcmc>\n",
    "\n",
    "\t<!-- START Marginal Likelihood Estimator                                     -->\n",
    "\n",
    "\t<!-- Define marginal likelihood estimator (PS/SS) settings                   -->\n",
    "\t<marginalLikelihoodEstimator chainLength=\"500000\" pathSteps=\"50\" pathScheme=\"betaquantile\" alpha=\"0.3\">\n",
    "\t\t<samplers>\n",
    "\t\t\t<mcmc idref=\"mcmc\"/>\n",
    "\t\t</samplers>\n",
    "\t\t<pathLikelihood id=\"pathLikelihood\">\n",
    "\t\t\t<source>\n",
    "\t\t\t\t<joint idref=\"joint\"/>\n",
    "\t\t\t</source>\n",
    "\t\t\t<destination>\n",
    "\t\t\t\t<prior idref=\"prior\"/>\n",
    "\t\t\t</destination>\n",
    "\t\t</pathLikelihood>\n",
    "\t\t<log id=\"MLELog\" logEvery=\"1000\" fileName=\"%s.mle.log\">\n",
    "\t\t\t<pathLikelihood idref=\"pathLikelihood\"/>\n",
    "\t\t</log>\n",
    "\t</marginalLikelihoodEstimator>\n",
    "\n",
    "\t<!-- Stepping-stone sampling estimator from collected samples                -->\n",
    "\t<steppingStoneSamplingAnalysis fileName=\"%s.mle.log\" resultsFileName=\"%s.mle.result.log\">\n",
    "\t\t<likelihoodColumn name=\"pathLikelihood.delta\"/>\n",
    "\t\t<thetaColumn name=\"pathLikelihood.theta\"/>\n",
    "\t</steppingStoneSamplingAnalysis>\n",
    "\n",
    "\t<!-- END Marginal Likelihood Estimator                                       -->\n",
    "\n",
    "\t<report>\n",
    "\t\t<property name=\"timer\">\n",
    "\t\t\t<mcmc idref=\"mcmc\"/>\n",
    "\t\t</property>\n",
    "\t</report>\n",
    "\t\n",
    "</beast>\n",
    "\"\"\" % (ds,ds,ds,ds,ds,ds))\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b6097b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
